<a href="https://colab.research.google.com/github/Argentan/DMA_LAB2/blob/master/tutoriales/02_intro_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Breve introducción a Python

Python es un lenguaje interpretado de tipos dinámicos. A diferencia de R, por ejemplo, es un lenguaje de uso general y su capacidad para las ciencias de datos no viene dada desde sus librerías internas sino por los módulos externos. Se puede usar en forma imperativa u orientada a objetos; teóricamente también dentro del paradigma de programación funcional, pero con ciertas limitaciones (existen límites para ejecución recursiva, por ejemplo).

Python puro tiende a ser lento (para ejecutar, rápido para escribir), por tener tipos dinámicos y por el GIL. Sin embargo, el intérprete más comúnmente usado (CPytyhon) esta implementado en C y permite la integración de módulos externos de C, C++ y Fortran que, bien utilizados, permiten la aceleración de las partes claves de un programa. 

Desde lo estético, Python reemplaza las llaves por indentaciones, cosa que fuerza un código más legible. Por lo demás, existe toda una "filosofía" de lo que es "pythonico", en términos de prolijidad, funcionalidad y belleza del código.

## import this

"import this" es un módulo incorporado en Python que imprime el poema "The Zen of Python" (El Zen de Python) en la consola. El poema fue escrito por Tim Peters, un destacado desarrollador de software de Python, y es una colección de principios que los programadores de Python pueden seguir para escribir código legible y elegante.

El poema contiene 19 líneas de código que resumen los principios de diseño y estilo de Python, como la legibilidad, la simplicidad y la explícitud. El uso de "import this" en Python es una forma divertida y educativa de recordar a los programadores la importancia de seguir buenas prácticas de programación al escribir código en Python.

In [1]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


## Operaciones Básicas


| Operador  | Operación                        | Ejemplo         |
| --------- | -------------------------------- | --------------- |
| **        | Potencia                         | `2 ** 3 = 8`    |
| %         | Resto                            | `22 % 8 = 6`    |
| //        | División de enteros              | `22 // 8 = 2`   |
| /         | División normal                  | `22 / 8 = 2.75` |
| *         | Multiplicación                   | `3 * 3 = 9`     |
| -         | Resta                            | `5 - 2 = 3`     |
| +         | Suma         | `2 + 2 = 4`     |

In [2]:
22 / 8

2.75

In [3]:
22 // 8

2

In [4]:
22 % 8

6

## Variables

Las variables se asignan con el signo igual =, y para definir el nombre solo se pueden usar letras, números y guion bajo (_); no se pueden usar espacios y tampoco pueden empezar con un número.

In [5]:
# 123var = 123
# A123var

In [6]:
var = 123

In [7]:
type(var)

int

In [8]:
var_float = 3.1415
var_float

3.1415

In [9]:
type(var_float)

float

## Tipos básicos de datos


| Tipo  | Ejemplo
| --------- | --------------- |
| Lógico    | True, False| 
| Entero    | 1, 2, 3| 
| Decimal o punto flotante         | 1.25, -120.25 | 
| Cadena o texto        | Hola, Adiós              | 
| Lista        | \[1, 2, 3\]              |
| Diccionario        | {"llave": "valor"}              |

In [10]:
cadena = 'esto es una cadena de texto'

entero = 10

punto_flotante = 10.4

punto_flotante_alternativa = 10.

booleano = True  #False

lista = ["elemento1", "elemento2", 1236, 10.2, 2., True, 'ultimo_elemento']

tupla = ("elemento1", "elemento2", 1236, 10.2, 2., True, 'ultimo_elemento')

diccionario = {"llave1": 1 , "llave2": 2}


Vemos las variables usando el comando mágico "whos"


In [11]:
%whos

Variable                     Type      Data/Info
------------------------------------------------
booleano                     bool      True
cadena                       str       esto es una cadena de texto
diccionario                  dict      n=2
entero                       int       10
lista                        list      n=7
punto_flotante               float     10.4
punto_flotante_alternativa   float     10.0
this                         module    <module 'this' from '/usr/lib/python3.9/this.py'>
tupla                        tuple     n=7
var                          int       123
var_float                    float     3.1415


### Ejercicios

1. Asignar a un objeto llamado "numero" el resultado de: 100 / 4

In [12]:
numero = 100 / 4
numero

25.0

2. Ver el tipo de dato del objeto numero

In [13]:
type(numero)

float

3. Pedir al usuario un número entero y luego imprima si el número es par o impar.
Pista: para pedir un número entero al usuario usar: `int(input("Num: ")`


In [14]:
numero = int(input("Ingrese un número entero: "))
numero

Ingrese un número entero: 3


3

Este código Python solicita al usuario que ingrese un número entero y lo almacena en la variable "numero".

La función "input()" se utiliza para solicitar al usuario que ingrese un valor en la consola, y la función "int()" se utiliza para convertir el valor ingresado en un número entero.

El mensaje "Ingrese un número entero: " se muestra en la consola para indicarle al usuario qué tipo de entrada se espera.

In [15]:
resto = numero % 2      # mod
resto

1

In [16]:
if resto == 0:
  print('el numero es par')
else:
  print('el numero es impar')

el numero es impar


4. Pedir al usuario su edad y luego determinar si es mayor de edad (18 años o más) o no.

In [17]:
edad = int(input("Ingrese su edad: "))

Ingrese su edad: 26


In [19]:
if edad >= 18:
   print('es mayor de edad')
else:
  print('no es mayor de edad')

es mayor de edad


4. Calcular el promedio de tres números ingresados por el usuario.

In [20]:
# Solicitamos al usuario ingresar los tres números y los almacenamos en variables separadas
numero1 = float(input("Ingrese el primer número: "))
numero2 = float(input("Ingrese el segundo número: "))
numero3 = float(input("Ingrese el tercer número: "))

# Calculamos el promedio de los tres números
promedio = (numero1 + numero2 + numero3) / 3

# Imprimimos el resultado del promedio
print("El promedio de los tres números es:", promedio)

Ingrese el primer número: 5
Ingrese el segundo número: 6
Ingrese el tercer número: 45
El promedio de los tres números es: 18.666666666666668


### Operaciones con cadenas de texto

Las cadenas se pueden sumar y multiplicar. Dependiendo el caso, multiplicar es razonable, pero sumar está desaconsejado. Se recomienda usar formateo de texto.

In [ ]:
palabra1 = "hola"
palabra2 = "mundo"
palabra3 = palabra1 + palabra2
palabra3

'holamundo'

Usando formateo de texto:

In [ ]:
f"{palabra1} {palabra2}"

'hola mundo'

La expresión f"{palabra1} {palabra2}" es una forma de formatear cadenas (strings) en Python usando una f-string (formatted string literal).

Las f-strings son una característica introducida en Python 3.6 que permiten incrustar valores de variables o expresiones en una cadena de texto de manera más sencilla y legible que las formas tradicionales de formateo de cadenas.

En el ejemplo dado, la f-string está compuesta por dos variables llamadas palabra1 y palabra2. La sintaxis {variable} dentro de la f-string indica que Python debe tomar el valor de la variable y colocarlo en el lugar donde se encuentra la expresión {variable} dentro de la cadena. De esta manera, la f-string combina los valores de las dos variables en una sola cadena de texto, separadas por un espacio.

In [22]:
# Formateo de números: número de dígitos decimales y separadores de miles.

num = 1233432.127
por = 0.1568

f"Separador de miles y decimales: {num:,.2f}, o porcentajes {por:.0%}"

'Separador de miles y decimales: 1,233,432.13, o porcentajes 16%'

En este código se utiliza una f-string (formatted string literal) para crear una cadena de texto (string) con valores de variables insertados en ella de forma formatada.

En el ejemplo, se están utilizando dos variables numéricas: num y por. En la f-string, se utilizan diferentes opciones de formateo para dar formato a estas variables dentro de la cadena de texto.

`{num:,.2f}`: el formato ,.2f especifica que se deben usar dos decimales y que se deben separar los millares con comas: 1,233,432.13.

`{por:.0%}`: el formato :.0% especifica que no se deben mostrar decimales y que el valor debe multiplicarse por 100 y mostrarse con un signo de porcentaje. En este caso 16%.

En resumen, la f-string permite combinar valores de variables con texto y aplicar opciones de formateo para darles un formato específico dentro de la cadena de texto. Esto puede ser muy útil para crear mensajes personalizados, mostrar resultados numéricos con un formato específico o presentar datos de forma legible para el usuario.

### Funciones básicas de las listas

Los elementos de una lista se pueden acceder mediante su índice

* Todos los índices empiezan en 0
* Se usan números enteros
* El índice máximo equivale a la cantidad de (elementos - 1)
* Se accede al elemento con [*índice*]
* Si el índice es negativo, se cuenta desde el final
* Se determina la longitud con la función `len`

In [23]:
lista

['elemento1', 'elemento2', 1236, 10.2, 2.0, True, 'ultimo_elemento']

In [24]:
n = len(lista)
n

7

In [25]:
lista[0]

'elemento1'

In [26]:
lista[n-1]

'ultimo_elemento'

In [27]:
tupla

('elemento1', 'elemento2', 1236, 10.2, 2.0, True, 'ultimo_elemento')

In [28]:
tupla[0]

'elemento1'

In [29]:
lista[-1]

'ultimo_elemento'

In [30]:
lista[-2]

True

In [31]:
lista[len(lista)-1]

'ultimo_elemento'

In [34]:
lista[0] = "otro elemento"
lista

['otro elemento', 'elemento2', 1236, 10.2, 2.0, True, 5]

In [35]:
tupla[0] = "otro elemento"

TypeError: ignored

Se recibe este mensaje de error porque no se pueden cambiar los valores dentro de una tupla.

## Seleccionando rangos
Se puede seleccionar más de un elemento de una lista usando '**:**'

Para seleccionar los primeros 2 se hace **[:2]**

###Slicing (rebanado)

Guido van Rossum:

[...] Me dejé llevar por la elegancia de los intervalos entreabiertos. Especialmente la invariante de que cuando dos rebanadas son adyacentes, el índice final de la primera rebanada es el índice de inicio de la segunda rebanada: esto es demasiado hermoso para ignorarlo. Por ejemplo, suponga que divide una cadena en tres partes en los índices i y j; las partes serían a[:i], a[i:j] y a[j:].


`a[start:stop]`  # desde start hasta stop-1

`a[start:]`      # desde start hasta el final

`a[:stop]`       # desde el comienzo hasta stop-1

`a[:]`           # copia de todo

`a[start:stop:step]` # desde start hasta stop, por pasos

La otra característica es que start y stop pueden ser negativos, lo que significa que cuenta desde el final. Entonces:

`a[-1]` # último elemento

`a[-2:]` # últimos dos elementos

`a[:-2]` # todo excepto los dos últimos elementos

In [36]:
lista

['otro elemento', 'elemento2', 1236, 10.2, 2.0, True, 5]

In [37]:
lista[-1]

5

In [38]:
lista[-2:]

[True, 5]

In [39]:
lista[:-2]

['otro elemento', 'elemento2', 1236, 10.2, 2.0]

In [40]:
lista[:2]

['otro elemento', 'elemento2']

Los últimos dos:

In [41]:
lista[-2:]

[True, 5]

Es lo mismo que seleccionar desde el 5to índice hasta el último índice

In [44]:
lista[5:]

[True, 5]

Del segundo al cuarto (ojo que el segundo elemento es el índice 1)

In [45]:
lista[1:4]

['elemento2', 1236, 10.2]

De manera similar, `step` puede ser un número negativo:

`a[::-1]` # todos los elementos, invertidos

`a[1::-1]` # los dos primeros elementos, invertidos

`a[:-3:-1]` # los dos últimos elementos, invertidos

`a[-3::-1]` # todo excepto los dos últimos elementos, al revés

In [46]:
lista

['otro elemento', 'elemento2', 1236, 10.2, 2.0, True, 5]

In [47]:
lista_invertida = lista[::-1]
lista_invertida

[5, True, 2.0, 10.2, 1236, 'elemento2', 'otro elemento']

In [48]:
lista[1::-1]

['elemento2', 'otro elemento']

In [ ]:
lista[:-3:-1]

[5, True]

In [ ]:
lista[-3::-1]

[2.0, 10.2, 1236, 'elemento2', 'otro elemento']

## Ejercicios

#### Listas

1. Crear una lista llamada "Letras" que contenga 5 letras de la A a la E  (A, B, C, D, E)

In [58]:
Letras = ['A', 'B', 'C', 'D', 'E']

2. Mostrar la longitud de la lista

In [59]:
len(Letras)

5

2. Mostrar, usando el índice, el primer y el último elemento de la lista

In [60]:
# Mostrar el primer elemento
print("El primer elemento es:", Letras[0])

# Mostrar el último elemento
print("El último elemento es:", Letras[len(Letras)-1])

El primer elemento es: A
El último elemento es: E


3. Mostrar, usando el índice, los tres primeros elementos de la lista

In [61]:
Letras[:3]

['A', 'B', 'C']

4. Mostrar Letras en orden inverso

In [62]:
Letras_inverso = Letras[::-1]
Letras_inverso

['E', 'D', 'C', 'B', 'A']

### Agregando datos a una Lista

Usando `append()` se agrega al final de la lista.

In [ ]:
nombres = ['Pedro', 'Lucia', 'Juan', 'Maria']
nombres.append('Manuel')
nombres

['Pedro', 'Lucia', 'Juan', 'Maria', 'Manuel']

Usando `insert()`: le digo la posición donde insertarlo.

In [ ]:
nombres = ['Pedro', 'Lucia', 'Juan', 'Maria']
nombres.insert(1, 'Pablo')
nombres

['Pedro', 'Pablo', 'Lucia', 'Juan', 'Maria']

In [ ]:
nombres = ['Pedro', 'Lucia', 'Juan', 'Maria']
nombres.insert(len(nombres), 'Mariana')
nombres

['Pedro', 'Lucia', 'Juan', 'Maria', 'Mariana']

### Eliminar valores de una lista y ordenar

Para eliminar usar `remove()`

In [ ]:
nombres = ['Pedro', 'Lucia', 'Juan', 'Maria', 'Lucia']
nombres.remove('Lucia')
nombres

['Pedro', 'Juan', 'Maria', 'Lucia']

Puedo ordenarlos usando `sort()`

In [ ]:
numeros = [2, 5, 3.14, 1, -7]
numeros.sort()
numeros

[-7, 1, 2, 3.14, 5]

In [ ]:
nombres = ['Pedro', 'JUAN', 'Juan', 'juan', 'Maria']
nombres.sort()
nombres

['JUAN', 'Juan', 'Maria', 'Pedro', 'juan']

Puedo invertir el orden de ordenamiento con `reverse=`

In [ ]:
numeros.sort(reverse=True)
numeros

[5, 3.14, 2, 1, -7]

### Ejercicios

Dada una lista de números del 1 al 10, realizar lo que se pide.

`numeros = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]`

* Agregar el número 11 al final de la lista.
* Eliminar el número 3 de la lista.
* Agregar el nùmero 5.5 entre el 2 y el 4
* Ordenar la lista en orden ascendente e imprimir
* Ordenar la lista en orden descendente e imprimir

In [63]:
numeros = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [64]:
numeros.append(11)
numeros

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [65]:
numeros.remove(3)
numeros

[1, 2, 4, 5, 6, 7, 8, 9, 10, 11]

In [66]:
numeros.insert(2, 5.5)
numeros

[1, 2, 5.5, 4, 5, 6, 7, 8, 9, 10, 11]

In [67]:
numeros.sort()
numeros

[1, 2, 4, 5, 5.5, 6, 7, 8, 9, 10, 11]

In [68]:
numeros.sort(reverse=True)
numeros

[11, 10, 9, 8, 7, 6, 5.5, 5, 4, 2, 1]

## Diccionarios - Funciones básicas

In [ ]:
diccionario

{'llave1': 1, 'llave2': 2}

In [ ]:
diccionario.keys()

dict_keys(['llave1', 'llave2'])

In [ ]:
diccionario.values()

dict_values([1, 2])

In [ ]:
diccionario.items()

dict_items([('llave1', 1), ('llave2', 2)])

In [ ]:
diccionario["llave1"]

1

### Ejercicios

1. Diccionario personas

* Crear un diccionario vacío llamado `personas`.
* Usando llaves `persona1`, `persona2`, etc., agregar dos personas al diccionario `personas` con los siguientes atributos: nombre, edad, ciudad.
* Agregar una tercera persona al diccionario.
* Imprimir la edad de la segunda persona.
* Imprimir la ciudad de la primera persona.
* Eliminar la tercera persona del diccionario.
* Imprimir el número de elementos que hay en el diccionario.
* Imprimir las llaves (claves) del diccionario
* Imprimir los valores del diccionario 

In [78]:
# Crear un diccionario vacío
personas = {}

In [79]:
personas["persona1"] = {"nombre": "Juan", "edad": 25, "ciudad": "CABA"}
personas["persona2"] = {"nombre": "Ana", "edad": 30.2, "ciudad": "Salta"}

In [80]:
personas

{'persona1': {'nombre': 'Juan', 'edad': 25, 'ciudad': 'CABA'},
 'persona2': {'nombre': 'Ana', 'edad': 30.2, 'ciudad': 'Salta'}}

In [81]:
# Agregar una tercera persona al diccionario.
personas["persona3"] = {"nombre": "Pedro", "edad": 40, "ciudad": "Rosario"}

In [82]:
# Imprimir la edad de la segunda persona.
print("La edad de la segunda persona es:", personas["persona2"]["edad"])

La edad de la segunda persona es: 30.2


In [83]:
# Imprimir la ciudad de la primera persona.
print("La ciudad de la primera persona es:", personas["persona1"]["ciudad"])

La ciudad de la primera persona es: CABA


In [84]:
# Eliminar la tercera persona del diccionario.
del personas["persona3"]

In [88]:
# Imprimir el número de elementos que hay en el diccionario.
print("El número de elementos en el diccionario es:", len(personas))

El número de elementos en el diccionario es: 2


In [87]:
# Imprimir las llaves (claves) del diccionario
print("Las llaves del diccionario son:", list(personas.keys()))

Las llaves del diccionario son: ['persona1', 'persona2']


In [86]:
# Imprimir los valores del diccionario
print("Los valores del diccionario son:", list(personas.values()))

Los valores del diccionario son: [{'nombre': 'Juan', 'edad': 25, 'ciudad': 'CABA'}, {'nombre': 'Ana', 'edad': 30.2, 'ciudad': 'Salta'}]
